### PR TITLE
Feature issue mumuki/mumuki platform#306 display inline errors

### PR DESCRIPTION
**Pull request by julian-berbel**
Related to mumuki/mumuki-platform#306

### CODE REVIEW

File: lib/mumukit/explainer.rb [hunk 1]
```diff
@@ -1,16 +1,21 @@
...
     explain_methods
-        .map { |selector, key| eval_explain(selector, key, content, test_results) }
+        .flat_map { |selector, key| eval_explain(selector, key, content, test_results) }
         .compact
-        .map { |explain| I18n.t(explain[:key], explain[:binding]) }
-        .map { |it| "* #{it}" }
-        .join("\n")
+        .map do |key, binding|
+          if binding.key?(:type)
+            explanation = binding.merge message: I18n.t(key), line: binding[:line].to_i
```
Comment (flbulgarelli) [inline lib/mumukit/explainer.rb]: Those lines are a little hard to understand. Could you please extract some methods to improve readability?

File: lib/mumukit/explainer.rb [hunk 2]
```diff
@@ -25,4 +18,19 @@ def explain_methods
...
+      merge_numeric_key explanation, binding, :line
+      merge_numeric_key explanation, binding, :column
+      explanation
+    else
+      "* #{I18n.t key, binding}"
+    end
+  end
+
+  def merge_numeric_key(explanation, binding, key)
+    explanation.merge! key binding[key].to_i if binding.key? key
```
Comment (flbulgarelli) [inline lib/mumukit/explainer.rb]: Is there a missing `=>` perhaps?

**Comment by flbulgarelli** [comment]
Good. Just an extra change request:

After some debate, we concluded we should not mix this new data - the `inline_feedback` -  with the normal feedback, in order to keep backward compatibilty on the `feedback`  field - both in the transport and persistence layer. 

That said, `test_server` component should return **both** feedback and `inline_feedback`, but in separate fields. Something like this:

```ruby
raw_feedback = feedback_hook.run! ....
if raw_feedback.is_a? String
{
  feedback: raw_feedback 
}
else
inline_feedback, feedback = raw_feedback.partition { |it| it[:type].present? }
{
   feedback: feedback, 
   inline_feedback: inline_feedback
}
end
```

File: lib/mumukit/server/test_pipeline.rb [hunk 3]
```diff
@@ -12,15 +12,18 @@ def evaluate!
   end
 
   def generate_feedback!
-    @feedback = @server.run_feedback! @request,
+    feedback = @server.run_feedback! @request,
                                      struct(test_results: @test_results, expectation_results: @expectation_results)
+    @inline_errors, @feedback = feedback.partition {|it| it.key? :type}
```
Comment (flbulgarelli) [inline lib/mumukit/server/test_pipeline.rb]: Please check if feedback is a string in order to keep compatibility like in https://github.com/mumuki/mumukit/pull/67#issuecomment-352778410

**Comment by julian-berbel** [comment]
After discussing this, we have decided to leave the feedback hook's explain methods outputs as a single hash instead of a list. When an explanation is an inline error, we can generate both the inline and bottom feedback.

We have also decided to reify the concept of explanation, possibly in ContextualizedExplanation and UncontextualizedExplanation.

**Pull request closed without merging**